js_modules/dagster-ui/packages/ui-core/src/assets/AssetEventMetadataEntriesTable.tsx

```diff
@@ -1,104 +1,202 @@
-      o.metadataEntries.map((entry) => ({timestamp: o.timestamp, runId: o.runId, entry})),
-    ),
+  // or if a metadata key is present on the definition and then emitted in an event,
+  // only show the latest version (first one found)
+  const allRows = uniqBy(
+    [...observationRows, ...eventRows, ...definitionRows]
+      .filter((row) => !filter || row.entry.label.toLowerCase().includes(filter.toLowerCase()))
+      .filter((row) => !(hideTableSchema && isTableSchemaEntry(row.entry))),
     (e) => e.entry.label,
   );
 
   return (
-    <AssetEventMetadataScrollContainer>
-      <AssetEventMetadataTable>
-        <tbody>
-          {metadataEntries.map((entry) => (
-            <tr key={`metadata-${entry.label}`}>
-              <td>
-                <Mono>{entry.label}</Mono>
-              </td>
-              <td>
-                <Mono>
-                  <MetadataEntry entry={entry} expandSmallValues={true} />
-                </Mono>
-              </td>
-              <td style={{opacity: 0.7}}>{entry.description}</td>
-            </tr>
-          ))}
-          {observationEntries.map((obv) => (
-            <tr key={`metadata-${obv.timestamp}-${obv.entry.label}`}>
-              <td>
-                <Mono>{obv.entry.label}</Mono>
-              </td>
-              <td>
-                <Mono>
-                  <MetadataEntry entry={obv.entry} expandSmallValues={true} />
-                </Mono>
-              </td>
-              <td style={{opacity: 0.7}}>
-                <Box>
-                  {`Observed in run `}
-                  <Link to={`/runs/${obv.runId}?timestamp=${timestamp}`}>
-                    <Mono>{titleForRun({id: obv.runId})}</Mono>
-                  </Link>
-                </Box>
-                <Caption>
-                  {`(${dayjs(Number(obv.timestamp)).from(
-                    Number(timestamp),
-                    true /* withoutSuffix */,
-                  )} later)`}
-                </Caption>
-                {obv.entry.description}
-              </td>
-            </tr>
-          ))}
-        </tbody>
-      </AssetEventMetadataTable>
-    </AssetEventMetadataScrollContainer>
+    <>
+      {showFilter && (
+        <Box padding={{bottom: 12}}>
+          <TextInput
+            value={filter}
+            style={{minWidth: 250}}
+            icon="search"
+            onChange={(e) => setFilter(e.target.value)}
+            placeholder="Filter metadata keys"
+          />
+        </Box>
+      )}
+      <AssetEventMetadataScrollContainer>
+        <StyledTableWithHeader>
+          {showHeader && (
+            <thead>
+              <tr>
+                <td>Key</td>
+                {showTimestamps && <td style={{width: 200}}>Timestamp</td>}
+                <td>Value</td>
+                {showDescriptions && <td>Description</td>}
+              </tr>
+            </thead>
+          )}
+          <tbody>
+            {allRows.length === 0 && (
```

Comment:
Should we virtualize this table? I know its a pain but maybe better safe than sorry? At some point somebody is going to make way too many metadata entries

Reply:
I talked to Josh about this and the other virtualization ones, and we decided to clip to a reasonable number of rows and show a "Show more" button, since it'd be awkward for the individual sections to scroll here. Updated all these accordingly!